Ported Fluent Theme to ControlThemes.

grokys opened the pull request:
## What does the pull request do?

This PR ports the fluent theme to use `ControlTheme` with @Takoooooo's help.

Due to the way that control themes work, some of the themes required a bit of refactoring - in particular we decided not to allow nested `/template/` selectors in control themes, so controls that used these had to have their sub-controls refactored out into separate sub-`ControlTheme`s. This is a bit of a pain but should hopefully improve the stability of themes in the long-term as control themes are no longer relying on implementation details of other control themes.

The only place the lack of nested `/template/` selectors turned out to really be a problem was in `SplitButton` as that control essentially adds additional states to `Button` (checked, flyout open). For the moment I've used a hack with setting the `Tag` property but we really need a proper solution for this - probably something like `ClassSetter` which would allow setting classes on controls from a style, or maybe something like the CSS `:part` selector (or both).

Depends on #7679 
Depends on #8263 

# IMPORTANT

Currently targeting #7679 - change the PR target back to master before merging.

grokys committed:
WIP: Porting fluent theme to use control themes.

Co-Authored-By: Takoooooo <tako0qq@gmail.com>

grokys committed:
Remove trailing /template/ selectors.

grokys committed:
WIP: HamburgerMenu.

grokys committed:
Added ItemContainerTheme property.

Otherwise there's no easy way to apply themes to item containers.

grokys committed:
Ported HamburgerMenu to ControlTheme.

grokys committed:
Don't set Foreground in PathIcon theme.

It is more useful if this property inherits by default.

grokys committed:
Ported ScrollBar to ControlTheme.

grokys committed:
Remove unneeded normal state.

grokys committed:
Fix Button/ToggleButton transitions.

grokys committed:
Ported ButtonSpinner to ControlTheme.

grokys committed:
Tweaked CheckBox ControlTheme.

grokys committed:
Validated ComboBox/Item ControlTheme.

grokys committed:
Formatted ContentControl ControlTheme.

grokys committed:
Formatted DataValidationErrors ControlTheme.

grokys committed:
Tidied up DropDownButton ControlTheme.

grokys committed:
Tidied up EmbeddableControlRoot ControlTheme.

grokys committed:
Tweaked Expander ControlTheme.

And added "Rounded" option to ControlCatalog for testing.

grokys committed:
Formated ItemsControl ControlTheme.

grokys committed:
Formatted Label ControlTheme.

grokys committed:
Formatted ListBox/Item ControlTheme.

grokys committed:
Formatted Menu/FlyoutPresenter ControlTheme.

grokys committed:
Tweaked NotificationCard ControlTheme.

grokys committed:
Formatted NumericUpDown ControlTheme.

grokys committed:
Formatted OverlayPopupHost ControlTheme.

grokys committed:
Formatted PathIcon ControlTheme.

grokys committed:
Formatted PopupRoot ControlTheme.

grokys committed:
Checked ProgressBar ControlTheme.

grokys committed:
Tweaked RepeatButton ControlTheme.

grokys committed:
Tweaked Slider ControlTheme.

grokys committed:
Formatted TabControl ControlTheme.

grokys committed:
Twaked TabItem ControlTheme.

grokys committed:
Tweaked TabStrip/Item ControlTemplates.

grokys committed:
Tweaked ToggleButton ControlTheme.

grokys committed:
Tweaked ToggleSwitch ControlTheme.

grokys committed:
Tweaked ToolTip ControlTheme.

grokys committed:
Formatted TransitioningContentControl ControlTheme.

grokys committed:
Added preview to TreeView ControlTheme.

grokys committed:
Formatted Window ControlTheme.

grokys committed:
Formatted WindowNotificationManager ControlTheme.

grokys committed:
Formatted Calendar ControlTheme.

grokys committed:
Formatted CalendarButton ControlTheme.

grokys committed:
Formatted CalendarDayButton ControlTheme.

grokys committed:
Tweaked CalendarItem ControlTheme.

grokys committed:
Formatted CalendarDatePicker ControlTheme.

grokys committed:
Formatted Carousel ControlTheme.

grokys committed:
Formatted ContextMenu ControlTheme.

grokys committed:
Ported TextBox to ControlTheme.

grokys committed:
Refactored CaptionButtons.

Converted buttons to be actual buttons instead of `Panel`s so that we get correct button behavior, i.e. the possibility to cancel by releasing the mouse outside the button.

grokys committed:
Reset button animations correctly.

Otherwise they keep on animating way after the interaction.

grokys committed:
Tweaked DatePicker ControlTheme.

grokys committed:
Ported ManagedFileChooser to ControlTheme.

grokys committed:
WIP: Menu/Item.

Something strange is happening when you open the menu a 3rd time.

grokys committed:
Formatted NativeMenuBar ControlTheme.

grokys committed:
Port ScrollViewer to ControlTheme.

Including splitting out the menu scroller into a separate xaml file for separate preview.

grokys committed:
Ported SplitView to ControlTheme.

grokys committed:
Ported TreeViewItem to ControlTheme.

grokys committed:
Tweaked TitleBar ControlTheme.

grokys committed:
Undo changed to TitleBar.

It was causing the title bar to be shown in all designer previews.

grokys committed:
Ported TimePicker to ControlTheme.

And refactored shared date/time picker components into a separate file.

grokys committed:
Ported SplitButton to ControlTheme.

Not happy with the Tag hack but no better way to add additional states to a control currently.

grokys committed:
Removed a few unneeded styles I missed.

And tweak some designer previews.

grokys committed:
Prevent ComboBox borders getting cut off.

grokys commented on src/Avalonia.Themes.Fluent/Controls/Slider.xaml:
```diff
@@ -1,275 +1,397 @@
+                  Name="grid"
+                  Margin="{TemplateBinding Padding}"
+                  RowDefinitions="Auto, *">
+                <ContentPresenter
```
While refactoring, I noticed that this `ContentPresenter` doesn't seem to be used. Was it part of a feature that was cut? Can it be removed from the template?

maxkatz6 replied:
Yes, it's for Header property that UWP has but we don't. I.e., TextBox.Header, Slider.Header.
It works similarly how our Label works, just built in the template.

grokys commented on src/Avalonia.Themes.Fluent/Controls/SplitButton.xaml:
```diff
@@ -1,37 +1,120 @@
+      <Setter Property="Foreground" Value="{DynamicResource SplitButtonForegroundPressed}" />
+    </Style>
+
+    <Style Selector="^[Tag=checked]">
```
This is the "tag hack" I mentioned in the description. I'm not happy about it, but it seems to work for now.

robloo replied:
Yes, this is very ugly... I hope the missing functionality is added so this doesn't become a permanent work-around.

maxkatz6 replied:
Just for case, this functionality wasn't removed. You still can use nested `/template/` selectors outside of ControlTheme.

robloo replied:
Well, as my luck goes, the places where it would be used (correctly) on my end are in ControlThemes. And I added a larger comment below but if Style still supports this it really seems like ControlTheme should as well.

grokys replied:
You can still actually host `<Style>`s in the root element of your control theme template in order to use nested template selectors, however I'd try to avoid it if possible.

grokys commented on src/Avalonia.Themes.Fluent/Controls/ToggleButton.xaml:
```diff
@@ -39,75 +45,79 @@
-  </Style>
+    <Style Selector="^:indeterminate">
+      <Style Selector="^ /template/ ContentPresenter#PART_ContentPresenter">
+        <Setter Property="Background" Value="{DynamicResource ToggleButtonBackgroundIndeterminate}" />
```
While testing I noticed that the indeterminate state for `ToggleButton` looks the same as the unchecked state. Is this intentional?

timunie replied:
This is an issue in the toggle switch as well. I had a discussion about this in MahApps back then, The solution was to disallow `null` for `IsChecked`-property. 

See also:
- https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/toggles
- https://www.nngroup.com/articles/toggle-switch-guidelines/

robloo replied:
Traditionally (in WPF/WinUI) CheckBox derives from ToggleButton and does make use of the three states. This is very common if a CheckBox is used at the top of a column for example representing each row selection.

Since other controls inherit from ToggleButton, and require three states, I don't think this can or should be removed. It is also just fine for ToggleButton to only consider two states in its style as it can only be toggled on/off by the user.

> While testing I noticed that the indeterminate state for ToggleButton looks the same as the unchecked state. Is this intentional?

Yes, this is intentional and matches WinUI. It is only really useful in derived controls like CheckBox.

<details>
  <summary>Source code from UWP Generic.xaml</summary>
  
```xaml
            <!-- Resources for Windows.UI.Xaml.Controls.Primitives.ToggleButton -->
            <StaticResource x:Key="ToggleButtonBackground" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundPointerOver" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundPressed" ResourceKey="SystemControlBackgroundBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundDisabled" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundChecked" ResourceKey="SystemControlHighlightAccentBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundCheckedPointerOver" ResourceKey="SystemControlHighlightAccentBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundCheckedPressed" ResourceKey="SystemControlHighlightBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundCheckedDisabled" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundIndeterminate" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundIndeterminatePointerOver" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundIndeterminatePressed" ResourceKey="SystemControlBackgroundBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBackgroundIndeterminateDisabled" ResourceKey="SystemControlBackgroundBaseLowBrush" />
            <StaticResource x:Key="ToggleButtonForeground" ResourceKey="SystemControlForegroundBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundPointerOver" ResourceKey="SystemControlHighlightBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundPressed" ResourceKey="SystemControlHighlightBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundDisabled" ResourceKey="SystemControlDisabledBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonForegroundChecked" ResourceKey="SystemControlHighlightAltChromeWhiteBrush" />
            <StaticResource x:Key="ToggleButtonForegroundCheckedPointerOver" ResourceKey="SystemControlHighlightAltChromeWhiteBrush" />
            <StaticResource x:Key="ToggleButtonForegroundCheckedPressed" ResourceKey="SystemControlHighlightAltChromeWhiteBrush" />
            <StaticResource x:Key="ToggleButtonForegroundCheckedDisabled" ResourceKey="SystemControlDisabledBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonForegroundIndeterminate" ResourceKey="SystemControlForegroundBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundIndeterminatePointerOver" ResourceKey="SystemControlHighlightBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundIndeterminatePressed" ResourceKey="SystemControlHighlightBaseHighBrush" />
            <StaticResource x:Key="ToggleButtonForegroundIndeterminateDisabled" ResourceKey="SystemControlDisabledBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrush" ResourceKey="SystemControlForegroundTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushPointerOver" ResourceKey="SystemControlHighlightBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushPressed" ResourceKey="SystemControlHighlightTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushDisabled" ResourceKey="SystemControlDisabledTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushChecked" ResourceKey="SystemControlHighlightAltTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushCheckedPointerOver" ResourceKey="SystemControlHighlightBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushCheckedPressed" ResourceKey="SystemControlTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushCheckedDisabled" ResourceKey="SystemControlDisabledTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushIndeterminate" ResourceKey="SystemControlForegroundTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushIndeterminatePointerOver" ResourceKey="SystemControlHighlightBaseMediumLowBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushIndeterminatePressed" ResourceKey="SystemControlHighlightTransparentBrush" />
            <StaticResource x:Key="ToggleButtonBorderBrushIndeterminateDisabled" ResourceKey="SystemControlDisabledTransparentBrush" />
```

</details>

grokys commented on src/Avalonia.Themes.Fluent/Controls/ComboBox.xaml:
```diff
@@ -125,7 +136,6 @@
                       Background="{DynamicResource ComboBoxDropDownBackground}"
                       BorderBrush="{DynamicResource ComboBoxDropDownBorderBrush}"
                       BorderThickness="{DynamicResource ComboBoxDropdownBorderThickness}"
-                      Margin="0,-1,0,-1"
```
I removed this line as it was causing the `ComboBox` popup border to be truncated at the top and bottom. I've noticed this for a long time but never investigated it. @danwalmsley you seem to have added this line, do you remember why and it is ok to remove it?

robloo replied:
That is a copy-paste from Fluent v1 in WinUI

https://github.com/microsoft/microsoft-ui-xaml/blob/7ed0bcf32a0e02b673a3d29555bd604d1f9ac0b0/dev/ComboBox/ComboBox_themeresources_v1.xaml#L721

That Margin (and any others like it) can safely be removed. It was just a work-around for WinUI/UWP specific issues.

grokys commented:
Rebased to credit @Takoooooo for the work he did on this.

grokys committed:
Fix ButtonSpinner borders.

grokys committed:
Use theme default CornerRadius.

When "Rounded" checkbox isn't checked.

grokys commented on src/Avalonia.Themes.Fluent/Controls/CaptionButtons.xaml:
```diff
@@ -1,72 +1,97 @@
-            <Viewbox>
-              <Path Stretch="UniformToFill" Fill="{TemplateBinding Foreground}" Data="M2048 1229v-205h-2048v205h2048z" />
+          </Button>
+          <Button x:Name="PART_MinimiseButton"
```
@danwalmsley I changed these to be buttons, because previously as panels you lost the ability to press and then cancel the press by releasing outside the bounds of the control.

grokys committed:
Make Separator look correct.

Previously there was a `MenuItem > Separator` selector, but I don't really see any reason to just apply this style directly to `Separator` itself. It's mainly used in menus and if one wants to use it elsewhere then they'd need to set these properties anyway.

robloo commented:
> In particular we decided not to allow nested /template/ selectors in control themes, so controls that used these had to have their sub-controls refactored out into separate sub-ControlThemes

I've found other places where directly accessing the template of a control within a template is very useful. Not trying to start another big discussion but if I had a vote I would not want to remove this functionality. It is totally find for one control default style to depend on another control's default style within the same theme (translating to Avalonia's "new terminology" default style = ControlTheme, Theme = FluentTheme).

Additionally, you are adding code duplication of styles so it would be easy for a button in one control style to be out-of-sync or outdated with the main button's default style. That is an undetectable error -- however, referring to a template part that no longer exists can be detected. So... well, I would not have done this for 

 1) Code duplication
 2) Sure there is separation now so some changes don't break dependent control styles; however, now you have undetectable places where code can become out-of-sync design-wise

grokys commented:
> I've found other places where directly accessing the template of a control within a template is very useful. Not trying to start another big discussion but if I had a vote I would not want to remove this functionality.

Yeah we had a pretty intense discussion about this internally as well, but the general consensus was:

- The `/template/` selector was generally speaking a mistake. It came from the CSS `/deep/` selector which was [removed](https://stackoverflow.com/questions/35741722/whats-the-substitute-for-shadow-and-deep) and for good reason - it's essentially the equivalent of changing private fields in a class using reflection
- A `/template/` selector which can poke into any template was needed previously because of the way our styles worked - there was no scoping of styles
- The CSS `/deep/` selector was replaced by the [`:part` selector](https://developer.mozilla.org/en-US/docs/Web/CSS/::part) which is much better because it defines an API contract

So the plan is to:

- Add a `:part` selector - the details of this need to be fleshed out, but we already have concept of "parts" and we already have the `[TemplatePart]` attribute, so conceptually it's a good fit for Avalonia
- Add a `ClassSetter` which can be used to add extra state to other controls from a style
- Long term, eventually do a gentle deprecation of the `/template/` selector outside control themes

To give some context: we're currently basically unable to change our templates between major releases because the template contents essentially become a public API due to `/template/`. So we have to balance the fact that 1) `/template` is a problem and 2) removing `/template/` would break a lot of code if we did it now.

Given the above we felt that preventing meddling with implementation details of other controls in a _new feature_ is a good start, but removing the feature as it already exists is probably a step too far right now.

We are actually open to removing the limitation on nested `/template/` selectors in control themes if a problem arises that literally cannot be solved by any other means, but as you know it's far easier to add features than remove them! So for the moment I'm hoping that by disallowing nested template selectors we can understand what exactly they're needed for and try to come up with better solutions for these use-cases.

Does that make sense?

grokys commented:
> A single theme can have inter-dependencies.

The problem with this is, how do you define a theme? Are we talking about the fluent theme as defined in `Avalonia.Themes.Fluent`? Or does it include `Avalonia.Controls.DataGrid`/`ColorPicker` fluent themes? If it also includes those then it probably should include 3rd party fluent themes, right? And so you're back to having the contents of our themes be a public API.

grokys commented:
Started a discussion to discuss the usages of `/template/` selectors: https://github.com/AvaloniaUI/Avalonia/discussions/8490

grokys commented on src/Avalonia.Themes.Fluent/Controls/PathIcon.xaml:
```diff
@@ -8,22 +8,18 @@
     </StackPanel>
   </Design.PreviewWith>
-  <Style Selector="PathIcon">
-    <Setter Property="Foreground" Value="{DynamicResource TextControlForeground}" />
```
I removed the `Foreground` setter here because it's really useful if `PathIcon` inherits the current foreground color instead of resetting it. Is that OK @maxkatz6 ?

maxkatz6 replied:
Yes, definitely.

grokys committed:
Fix ButtonSpinner disabled icon color.

maxkatz6 committed:
Remove duplicated ItemsPanel on hamburger menu

grokys committed:
Prevent crash when switching themes.

grokys committed:
Merge branch 'master' into feature/7120-fluent-control-themes

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022401-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
